File: conformance/cpp/async_demo.cpp
// Conformance consumer: async-demo sample, C++ target.
//
// Drives the std::future-backed async surface of the generated header-only
// wrapper end to end: `weaveffi::tasks::run_task` settled from the producer's
// worker thread with a TaskResult struct decoded from a value buffer, the
// typed InvalidNameError (extending TaskError extending WeaveFFIError)
// rethrown by future::get for an empty name, the buffered list-of-records
// round trip through `run_batch`, the direct-scalar `run_n_tasks`, the sync
// `cancel_task`, and `active_callbacks` settling to zero once every task body
// has completed. Exits 0 on success; aborts (non-zero) on any failed
// assertion.

#include <cassert>
#include <cstdio>
#include <cstring>
#include <string>
#include <vector>

#include "weaveffi.hpp"

int main() {
    // Async record return: future::get blocks until the worker-thread
    // callback delivers the decoded TaskResult.
    weaveffi::TaskResult result = weaveffi::tasks::run_task("alpha").get();
    assert(result.id > 0);
    assert(result.value == "completed: alpha");
    assert(result.success);

    // Typed async error: the empty name settles the future with the typed
    // exception, rethrown on get().
    bool threw = false;
    try {
        weaveffi::tasks::run_task("").get();
    } catch (const weaveffi::InvalidNameError& e) {
        threw = true;
        assert(e.code() == 1);
        assert(dynamic_cast<const weaveffi::TaskError*>(&e) != nullptr);
        assert(dynamic_cast<const weaveffi::WeaveFFIError*>(&e) != nullptr);
    }
    assert(threw && "expected InvalidNameError for empty name");

    // Buffered list-of-records both ways.
    std::vector<weaveffi::TaskResult> batch =
        weaveffi::tasks::run_batch({"a", "b", "c"}).get();
    assert(batch.size() == 3);
    const char* expected[3] = {"completed: a", "completed: b", "completed: c"};
    for (size_t i = 0; i < 3; i++) {
        assert(batch[i].value == expected[i]);
        assert(batch[i].success);
    }

    // Direct scalar through the async callback.
    assert(weaveffi::tasks::run_n_tasks(7).get() == 7);

    // Sync functions beside the async ones.
    assert(!weaveffi::tasks::cancel_task(1));

    // Every spawned task body has completed by the time its callback fires.
    assert(weaveffi::tasks::active_callbacks() == 0);

    std::printf("cpp async-demo conformance: OK\n");
    return 0;
}

File: conformance/swift/async_demo.swift
// Conformance consumer: async-demo sample, Swift target.
//
// Drives the CheckedContinuation-backed async surface end to end:
// `Tasks.runTask` as a throwing async static resumed from the producer's
// worker thread and decoded from a value buffer into the plain `TaskResult`
// struct, the typed `TaskError.invalidName` case (with its stable
// `errorCode`) thrown for an empty name, the buffered list-of-records round
// trip through the non-throwing async `runBatch`, the direct-scalar
// `runNTasks`, the sync `cancelTask`, and `activeCallbacks` settling to zero
// once every task body has completed.

import Foundation
import AsyncDemo

func fail(_ msg: String) -> Never {
    FileHandle.standardError.write(Data("assertion failed: \(msg)\n".utf8))
    exit(1)
}

func expect(_ cond: Bool, _ msg: String) {
    if !cond { fail(msg) }
}

// Async record return: resumed with a TaskResult decoded from the
// completion callback's value buffer.
do {
    let result = try await Tasks.runTask(name: "alpha")
    expect(result.id > 0, "runTask assigns an id")
    expect(result.value == "completed: alpha", "runTask value (got \(result.value))")
    expect(result.success, "runTask success flag")
} catch {
    fail("runTask threw unexpectedly: \(error)")
}

// Typed async error: the empty name resumes by throwing the typed case.
do {
    _ = try await Tasks.runTask(name: "")
    fail("expected TaskError.invalidName for empty name")
} catch let e as TaskError {
    guard case .invalidName = e else {
        fail("expected .invalidName (got \(e))")
    }
    expect(e.errorCode == 1, "invalidName carries code 1 (got \(e.errorCode))")
} catch {
    fail("expected TaskError (got \(error))")
}

// Buffered list-of-records both ways through the non-throwing async.
let batch = await Tasks.runBatch(names: ["a", "b", "c"])
expect(
    batch.map { $0.value } == ["completed: a", "completed: b", "completed: c"],
    "runBatch values"
)
expect(batch.allSatisfy { $0.success }, "runBatch success flags")

// Direct scalar through the async callback.
let n = await Tasks.runNTasks(n: 7)
expect(n == 7, "runNTasks echoes n")

// Sync functions beside the async ones.
expect(!Tasks.cancelTask(id: 1), "cancelTask reports not cancelled")

// Every spawned task body has completed by the time its callback fires.
expect(Tasks.activeCallbacks() == 0, "activeCallbacks settles to zero")

print("swift async-demo conformance: OK")
